tests/lobpcg_test.py

```diff
@@ -0,0 +1,238 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Tests for lobpcg routine."""
+
+import re
+import os
+
+from absl.testing import parameterized
+import numpy as np
+import scipy.linalg as sla
+
+from jax._src import test_util as jtu
+from jax._src.config import flags
+from jax.experimental.sparse import linalg
+import jax.numpy as jnp
+
+FLAGS = flags.FLAGS
+flags.DEFINE_string(
+    'lobpcg_debug_plot_dir',
+    os.getenv('LOBPCG_DEBUG_PLOT_DIR'),
+    help=
+    'If set, emits debug plots in given directory (requires matplotlib).'
+)
+
+def _make_lobpcg_cases():
+  example_names = list(_example_generators())
+  cases = []
+  for name in example_names:
+    nkm = [(100, 10, 20)]
+    if not FLAGS.jax_skip_slow_tests:
+      nkm.append((1000, 100, 200))
+    for n, k, m in nkm:
+      if name == 'ring laplacian' or name.startswith('linear'):
+        m *= 2
+      case = [('matrix_name', name), ('n', n), ('k', k), ('m', m)]
+      clean_matrix_name = re.sub('[^0-9a-zA-Z]+', '_', name)
+      case.append(('testcase_name', f'{clean_matrix_name}_n{n}'))
+      cases.append(dict(case))
+
+  assert len({c['testcase_name'] for c in cases}) == len(cases)
+  return cases
+
+def _make_ring(n, k):
+  # from lobpcg scipy tests
+  col = np.zeros(n)
+  col[1] = 1
+  A = sla.toeplitz(col)
+  D = np.diag(A.sum(axis=1))
+  L = D - A
+  # Compute the full eigendecomposition using tricks, e.g.
+  # http://www.cs.yale.edu/homes/spielman/561/2009/lect02-09.pdf
+  tmp = np.pi * np.arange(n) / n
+  analytic_w = 2 * (1 - np.cos(tmp))
+  analytic_w.sort()
+  analytic_w = analytic_w[::-1]
+
+  return L.astype(np.float32), analytic_w.astype(np.float32)
+
+def _make_diag(diag):
+  diag.sort()
+  diag = diag[::-1]
+  return np.diag(diag).astype(np.float32), diag.astype(np.float32)
+
+def _make_cluster(to_cluster, n):
+  return _make_diag(
+      np.array([1000] * to_cluster + [1] * (n - to_cluster)))
+
+def _example_generators():
+  return {
+      'id': lambda n, _k: _make_diag(np.ones(n)),
+      'linear cond=1k': lambda n, _k: _make_diag(np.linspace(1, 1000, n)),
+      'linear cond=100k':
+          lambda n, _k: _make_diag(np.linspace(1, 100 * 1000, n)),
+      'geom cond=1k': lambda n, _k: _make_diag(np.logspace(0, 3, n)),
+      'geom cond=100k': lambda n, _k: _make_diag(np.logspace(0, 5, n)),
+      'ring laplacian': lambda n, k: _make_ring(n, k),
+      'cluster(k/2)': lambda n, k: _make_cluster(k // 2, n),
+      'cluster(k-1)': lambda n, k: _make_cluster(k - 1, n),
+      'cluster(k)': lambda n, k: _make_cluster(k, n)}
+
+class LobpcgTest(jtu.JaxTestCase):
+  # Override base class default config.
+  _default_config = {
+      'jax_enable_checks': True,
+      'jax_debug_nans': True,
+      'jax_numpy_rank_promotion': 'raise',
+      'jax_traceback_filtering': 'off',
+  }
+
+  # TODO(vladf): add f64 tests just to verify it compiles?
```

Comment:
I think that would be good -- you could probably check if it's running on CPU and disable otherwise?

Reply:
It looks like there are `f64` tests auto-triggered by the github action matrix, so I think all I should  really need to do here is adjust the epsilons.